pkg/cluster/agent_customization.go

```diff
@@ -0,0 +1,86 @@
+		return cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetClusterAgentAffinity returns node affinity for the cluster agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetClusterAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	return unmarshalAffinity(settings.ClusterAgentDefaultAffinity.Get())
+}
+
+// GetClusterAgentResourceRequirements returns resource requirements (cpu, memory) for the cluster agent if it has been
+// user defined. If not, nil is returned.
+func GetClusterAgentResourceRequirements(cluster *v3.Cluster) *corev1.ResourceRequirements {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideResourceRequirements != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideResourceRequirements
+	}
+
+	return nil
+}
+
+// GetFleetAgentTolerations returns additional tolerations for the fleet agent if it has been user defined. If not,
+// then nil is returned.
+func GetFleetAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetFleetAgentAffinity returns node affinity for the fleet agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetFleetAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	return unmarshalAffinity(settings.FleetAgentDefaultAffinity.Get())
+}
+
+// GetFleetAgentResourceRequirements returns resource requirements (cpu, memory) for the fleet agent if it has been
+// user defined. If not, nil is returned.
+func GetFleetAgentResourceRequirements(cluster *v3.Cluster) *corev1.ResourceRequirements {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.OverrideResourceRequirements != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.OverrideResourceRequirements
+	}
+
+	return nil
+}
+
+// ummarshalAffinity returns an unmarshalled v1 node affinity object.
+func unmarshalAffinity(affinity string) *corev1.Affinity {
+	var affinityObj corev1.Affinity
+	err := json.Unmarshal([]byte(affinity), &affinityObj)
+	if err != nil {
+		logrus.Errorf("failed to unmarshal node affinity: %s", err)
```

Comment:
https://stackoverflow.com/a/61287626
```suggestion
		logrus.Errorf("failed to unmarshal node affinity: %w", err)
```